libcontainer/capabilities/capabilities.go

```diff
@@ -57,27 +58,47 @@ func New(capConfig *configs.Capabilities) (*Caps, error) {
 
 	cm := capMap()
 	unknownCaps := make(map[string]struct{})
+	ignoredCaps := make(map[string]struct{})
 	// capSlice converts the slice of capability names in caps, to their numeric
 	// equivalent, and returns them as a slice. Unknown or unavailable capabilities
 	// are not returned, but appended to unknownCaps.
-	capSlice := func(caps []string) []capability.Cap {
+	//
+	// If mustHave argument is not nil, caps that are not present in mustHave
+	// are appended to ignoredCaps instead of the resulting slice.
+	capSlice := func(caps []string, mustHave []capability.Cap) []capability.Cap {
```

Comment:
A "global" list of ignoredCaps that can be modified by any call to this function doesn't play nice with the fact that we only want these capabilities when we call this with ambient/inheritable. All the other calls, we don't want them to modify this.

Right now all the other calls use nil as the second param and only ambient/inheritable use non nil for the second param, so this works fine. But it seems a little error-prone.

What if it returns two caps set, the current value and the slice of caps not present in mustHave? Then only the call for ambient/inh will not ignore it? Or maybe we don't want to modify this capSlice func, maybe we want another function and use the slices package, like https://pkg.go.dev/slices#Compare and maybe delete, to get the ignored caps? 

I understand that the mustHave we probably only want it for ambient/inh now, so it should be safe now. But if the userns capabilities are merged upstream, or we do some change, this might cause some bug. I'm not sure the risk is big and worth changing this, but I think the alternatives I proposed seems also easier to reason about. 

What do you think?

Reply:
At this point, I think it's best to document the `capSlice` function making it clear it modifies some variables, letting future devs know about it.

I just tried doing the above and realized it is already documented ("caps ... are appended to ignoredCaps").

Since this function is scoped inside `New`, its usage is quite limited. I understand the argument for being future proof, but it looks like adding more parameters (and, especially, return values) will decrease readability, which plays against us in "future proof" department.

A separate code to check that all ambient capabilities are also present in inheritable will be the best way, I guess. Using slices.Compare is not an option here, because:
 - our slices are not sorted;
 - inheritable set can contain capabilities that ambient set does not.
 
So we have to have a `for` (or something like `slices.DeleteFunc`). The problem here is we either compare a bunch of strings (which is inefficient) or deal with resulting slices of ints, but then we have to convert those capabilities back to string for the warning message (which is OK since the warning is rare, but it adds yet another complexity).

I spent some time trying to come up with something that sucks less than the code I have here, but I fail miserably. Perhaps you can suggest some code @rata?

For now, I just added cleaning the ignoredCaps map after use, hoping that if someone is going to reuse this, they will copy-paste the code 😁. Here's the diff for your review:

```diff
 diff --git a/libcontainer/capabilities/capabilities.go b/libcontainer/capabilities/capabilities.go
index 1111aac3..8082b4f0 100644
--- a/libcontainer/capabilities/capabilities.go
+++ b/libcontainer/capabilities/capabilities.go
@@ -89,6 +89,7 @@ func New(capConfig *configs.Capabilities) (*Caps, error) {
        ambient := capSlice(capConfig.Ambient, inheritable)
        if len(ignoredCaps) > 0 {
                logrus.Warn("unable to set Ambient capabilities which are not set in Inheritable; ignoring following Ambient capabilities: ", mapKeys(ignoredCaps))
+               clear(ignoredCaps)
        }
 
        c.caps = map[capability.CapType][]capability.Cap{
```

Reply:
True, it is not sorted so it would be more expensive, and it is not very expensive (n^2 in 15 elem lists is not too much), but it would imply allocations and so on.

Okay, I think the current comment is fine :-)